src/framework/step/mod.rs

```diff
@@ -0,0 +1,132 @@
+use crate::framework::Error;
+#[cfg(feature = "enable-serde")]
+use serde::{
+    de::{Error as SerdeError, Visitor},
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::fmt::Formatter;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Step {
+    Example(ExampleCircuit),
+    ModulusConversion(ModulusConversionDetails),
+}
+
+impl Step {
+    #[must_use]
+    pub fn to_path(&self) -> String {
+        match self {
+            Self::Example(circuit) => format!("example/{}", circuit.to_path()),
+            Self::ModulusConversion(details) => format!("modulus_conversion/{}", details.to_path()),
+        }
+    }
+
+    pub fn from_path(path_str: &str) -> Result<Step, Error> {
+        let path_str = path_str.strip_prefix('/').unwrap_or(path_str);
+        let (step, next) = path_str.split_once('/').ok_or(Error::StepParseError)?;
+        match step {
+            "example" => Ok(Self::Example(ExampleCircuit::from_path(next)?)),
+            "modulus_conversion" => Ok(Self::ModulusConversion(
+                ModulusConversionDetails::from_path(next)?,
+            )),
+            _ => Err(Error::StepParseError),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ExampleCircuit {
+    Mul1(),
+    Mul2(),
+}
+
+impl ExampleCircuit {
+    #[must_use]
+    pub fn to_path(&self) -> String {
+        match self {
+            Self::Mul1() => "mul1".to_string(),
+            Self::Mul2() => "mul2".to_string(),
+        }
+    }
+
+    pub fn from_path(path_str: &str) -> Result<ExampleCircuit, Error> {
+        match path_str {
+            "mul1" => Ok(ExampleCircuit::Mul1()),
+            "mul2" => Ok(ExampleCircuit::Mul2()),
+            _ => Err(Error::StepParseError),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ModulusConversionDetails {
+    Mul1 { bit_num: u32 },
+    Mul2 { bit_num: u32 },
```

Comment:
```suggestion
    Mul1 { bit_num: u8 },
    Mul2 { bit_num: u8 },
```

Reply:
or even

```rust
   BitMul1(u8),
   BitMul2(u8)
```

we would need some documentation here anyway to explain what these operations are

Reply:
agreed. this isn't actually implemented yet, and honestly might be better to give these more descriptive names as well